kong/acl_service.go

```diff
@@ -94,67 +98,71 @@ func (s *ACLService) Delete(ctx context.Context,
-		if err != nil {
-			return nil, nil, err
-		}
-		var aclGroup ACLGroup
-		err = json.Unmarshal(b, &aclGroup)
-		if err != nil {
-			return nil, nil, err
-		}
-		aclGroups = append(aclGroups, &aclGroup)
-	}
-
-	return aclGroups, next, nil
+	return s.listByEndpointAndOpt(ctx, "/acls", opt)
 }
 
-// ListAll fetches all all ACL group associations in Kong.
+// ListAll fetches all ACL group associations in Kong.
 // This method can take a while if there
 // a lot of ACLGroup associations are present.
 func (s *ACLService) ListAll(ctx context.Context) ([]*ACLGroup, error) {
-	var aclGroups, data []*ACLGroup
-	var err error
-	opt := &ListOpt{Size: pageSize}
+	return s.ListAllByTags(ctx, nil)
+}
 
-	for opt != nil {
-		data, opt, err = s.List(ctx, opt)
-		if err != nil {
-			return nil, err
-		}
-		aclGroups = append(aclGroups, data...)
-	}
-	return aclGroups, nil
+// ListAllByTags fetches all ACL group associations filtered by tags in Kong.
+// This method can take a while if there
+// a lot of ACLGroup associations are present.
+func (s *ACLService) ListAllByTags(ctx context.Context, tags []string) ([]*ACLGroup, error) {
+	return s.ListAllByOpt(ctx, newOpt(tags))
+}
+
+// ListAllByOpt fetches all ACL group associations filtered by opt in Kong.
+// This method can take a while if there
+// a lot of ACLGroup associations are present.
+func (s *ACLService) ListAllByOpt(ctx context.Context, opt *ListOpt) ([]*ACLGroup, error) {
+	return s.listAllByEndpointAndOpt(ctx, "/acls", opt)
 }
 
 // ListForConsumer fetches a list of ACL groups
 // in Kong associated with a specific consumer.
 // opt can be used to control pagination.
 func (s *ACLService) ListForConsumer(ctx context.Context,
 	consumerUsernameOrID *string, opt *ListOpt) ([]*ACLGroup, *ListOpt, error) {
-	data, next, err := s.client.list(ctx,
-		"/consumers/"+*consumerUsernameOrID+"/acls", opt)
+	return s.listByEndpointAndOpt(ctx, "/consumers/"+*consumerUsernameOrID+"/acls", opt)
+}
+
+func (s *ACLService) listAllByEndpointAndOpt(ctx context.Context,
+	endpoint string, opt *ListOpt) ([]*ACLGroup, error) {
+	data, err := s.client.listAll(ctx, endpoint, opt, true)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	var aclGroups []*ACLGroup
 	for _, object := range data {
-		b, err := object.MarshalJSON()
+		var aclGroup ACLGroup
+		err = json.Unmarshal(object, &aclGroup)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
+		aclGroups = append(aclGroups, &aclGroup)
+	}
+
+	return aclGroups, nil
+}
+
+func (s *ACLService) listByEndpointAndOpt(ctx context.Context,
+	endpoint string, opt *ListOpt) ([]*ACLGroup, *ListOpt, error) {
+	data, next, err := s.client.list(ctx, endpoint, opt)
+	if err != nil {
+		return nil, nil, err
+	}
+	var aclGroups []*ACLGroup
+	for _, object := range data {
 		var aclGroup ACLGroup
-		err = json.Unmarshal(b, &aclGroup)
+		err = json.Unmarshal(object, &aclGroup)
 		if err != nil {
 			return nil, nil, err
 		}
```

Comment:
I changed data from `[]*json.RawMessage` to `[][]byte` to avoid `b, err := object.MarshalJSON()` being repeated everywhere so, would the following snippet do the same thing? 
```
func asACLGroups(in [][] byte ) ([]*ACLGroup, error) {
  // ...
}
```

Reply:
Done with 476c04b